### DISCUSSION TITLE
Feature Request: python: Flag to turn off peeking available variables

### DISCUSSION THREAD

**Discussion by NickCrews** [duckdb/duckdb#8100] [Ideas]
The ability to peek at available variables in the python stackframe

```python
import duckdb
r1 = duckdb.sql('SELECT 42 AS i')
duckdb.sql('SELECT i * 2 AS k FROM r1').show()
```

is very handy for interactive work. But when duckdb is extended, eg with Ibis, this can lead to some footguns because duckdb can find and use local variables that I would not have expected.

```python
import ibis
import pandas as pd

ibis.options.interactive = True

be2 = ibis.duckdb.connect()
be1 = ibis.duckdb.connect()
assert be1 is not be2

df1 = pd.DataFrame({"a": [0, 1, 2]})
df2 = pd.DataFrame({"a": [3, 4, 5]})
t1 = be1.create_table("a", df1)
t2 = be2.create_table("b", df2)

u = ibis.union(t1, t2)
try:
    u.execute() # This fails, as expected, because we are using two different connections
except Exception as e:
    print(e)
# If I bring a variable "b" into scope though, duckdb picks up the wrong data:
b = df1
u.execute()
```

I'm not exactly sure how the API would work, but can we add some flag to turn off this behavior?

Originally filed with Ibis at https://github.com/ibis-project/ibis/issues/6551

**Comment by Tishj** [comment]
I think that's a legitimate issue,
I hacked around for 20 minutes and came up with:
```py

    def test_disabled_local(self):
        con = duckdb.connect()
        df = pd.DataFrame({'A': [42]})
        con.options.scan_variables = False
        with pytest.raises(duckdb.CatalogException):
            # Because we disabled scanning variables, no replacement will be done for 'df'
            con.sql('select * from df')
```

It might be a while until I get around to PRing this upstream though
here is the PR I made to my local fork which I made so I don't lose track of the branch
<https://github.com/Tishj/duckdb/pull/40>

**Reply by NickCrews** [reply]
Thanks for digging into this @Tishj!

@cpcloud as a temp/permanent workaround could/should we add this `con.options.scan_variables = False` config step every time a duckdb backend is created in ibis?
